core/google/api/core/gapic_v1/method.py

```diff
@@ -0,0 +1,222 @@
+
+
+def _prepare_metadata(metadata):
+    """Transforms metadata to gRPC format and adds global metrics.
+
+    Args:
+        metadata (Optional[Mapping[str, str]]): Any current metadata.
+
+    Returns:
+        Sequence[Tuple(str, str)]: The gRPC-friendly metadata keys and values.
+    """
+    if metadata is None:
+        metadata = {}
+
+    client_metadata = 'api-core/{} gl-python/{} grpc/{}'.format(
+        _API_CORE_VERSION, _PY_VERSION, _API_CORE_VERSION)
+
+    # Merge this with any existing metric metadata.
+    if METRICS_METADATA_KEY in metadata:
+        client_metadata = '{} {}'.format(
+            client_metadata, metadata[METRICS_METADATA_KEY])
+
+    metadata[METRICS_METADATA_KEY] = client_metadata
+
+    return list(metadata.items())
+
+
+class _GapicCallable(object):
+    """Callable that applies retry, timeout, and metadata logic."""
+
+    def __init__(self, target, retry, timeout, metadata):
+        """
+        Args:
+            target (Callable): The low-level RPC method. If timeout is not
+                False, should accept a timeout argument. If metadata is not
+                False, it should accept a metadata argument.
+            retry (google.api.core.retry.Retry): The default retry for the
+                callable. If False, this callable will not retry.
+            timeout (google.api.core.timeout.Timeout): The default timeout
+                for the callable. If False, this callable will not specify
+                a timeout argument to the low-level RPC method.
+            metadata (Sequence[Tuple[str, str]]): gRPC call metadata that's
+                passed to the low-level RPC method. If False, this callable
+                will not specify a metadata argument to the method.
+        """
+        self._target = target
+        self._retry = retry
+        self._timeout = timeout
+        self._metadata = metadata
+
+    def __call__(self, *args, **kwargs):
+        """Invoke the low-level RPC with retry, timeout, and metadata."""
+        # Note: Due to Python 2 lacking keyword-only arguments we use this
+        # pattern to extract the retry and timeout params.
+        retry_ = kwargs.pop('retry', None)
+        if retry_ is None:
+            retry_ = self._retry
+
+        timeout_ = kwargs.pop('timeout', None)
+        if timeout_ is None:
+            timeout_ = self._timeout
+
+        # If timeout is specified as a number instead of a Timeout instance,
+        # convert it to a ConstantTimeout.
+        if timeout_ is not False and isinstance(timeout_, (int, float)):
+            timeout_ = timeout.ConstantTimeout(timeout_)
+
+        # If timeout is the default and the default timeout is exponential and
+        # a non-default retry is specified, make sure the timeout's deadline
+        # matches the retry's. This handles the case where the user leaves
+        # the timeout default but specifies a lower deadline via the retry.
+        if (timeout_ is self._timeout
```

Comment:
I'd be more comfortable with a flag like `original_timeout` or something here rather than the identity check. Maybe I'm being paranoid?